### DISCUSSION TITLE
TopLevel.StorageProvider

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#9004] [Q&A]
Great, Avalonia 11 completely breaks [MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/).

For `AvaloniaSaveFileDialog`, it says

    CS0618: Class 'Avalonia.Controls.SaveFileDialog' is obsolete: Use Window.StorageProvider API or TopLevel.StorageProvider API

I can find absolutely no documentation about StorageProvider, and cannot find such class.

Can I get some information about it?

**Comment by maxkatz6** [comment]
> I can find absolutely no documentation about StorageProvider

Not yet. Avalonia 11 is only in the first preview. 
But documentation of the IStorageProvider interface and related classes is quite detailed. 

Call it from the window or toplevel (in case of mobile/web):
`window.StorageProvider.ShowOpenFilePicker()...`

**Comment by kekekeks** [comment]
>completely breaks

You can disable "warnings as errors", you know

**Comment by mysteryx93** [comment]
ok it's on window instance not on on the static type, found it.

Is this new Avalonia API considered stable? These are very drastic changes.

It's a little bit annoying that to support it, I'll need to redesign the whole API even for WPF applications that don't need it...

**Reply by maxkatz6** [reply]
Old API wasn't compatible with mobile and browser. The reason is simple.

**Reply by mysteryx93** [reply]
Totally understand that.

The goal of MvvmDialogs having standard API is that the ViewModel doesn't care whether you're running Avalonia or WPF or something else... but that API indeed doesn't support mobile and browser.

But then if I update the MvvmDialogs API to reflect the Avalonia API,

1. It will be very annoying for WPF users
2. Another platform could come along with yet a different API
3. It will be a major breaking change for existing users... Avalonia users expect the change but WPF users don't

**Reply by mysteryx93** [reply]
To keep in mind; what API is MvvmDialogs for MAUI going to use?

If someone knows the answer to that; is it closer to the new Avalonia API or closer to old WPF?

**Reply by kekekeks** [reply]
We still keep the old API as a compatibility layer which will continue to work.
For desktop platforms nothing has changed, if you want to be compatible with mobile platforms you need to use the new API.

**Reply by mysteryx93** [reply]
There is something else causing weird "invalid runtime" error; I'll see what it is after dealing with those errors first. Currently-published NuGet library doesn't work with v11.
